Core/include/Acts/TrackFitting/KalmanFitter.hpp

```diff
@@ -646,32 +651,31 @@ class KalmanFitter {
         // the lastTrackIndex.
         result.lastMeasurementIndex = result.lastTrackIndex;
 
-      } else if (surface->associatedDetectorElement() != nullptr ||
-                 surface->surfaceMaterial() != nullptr) {
+      } else if ((precedingMeasurementExists && surfaceIsSensitive) ||
+                 surfaceHasMaterial) {
         // We only create track states here if there is already measurement
         // detected or if the surface has material (no holes before the first
         // measurement)
-        if (result.measurementStates > 0 ||
-            surface->surfaceMaterial() != nullptr) {
-          auto trackStateProxyRes = detail::kalmanHandleNoMeasurement(
-              state, stepper, *surface, *result.fittedStates,
-              result.lastTrackIndex, true, logger(), freeToBoundCorrection);
-
-          if (!trackStateProxyRes.ok()) {
-            return trackStateProxyRes.error();
-          }
+        auto trackStateProxyRes = detail::kalmanHandleNoMeasurement(
+            state, stepper, *surface, *result.fittedStates,
+            result.lastTrackIndex, true, logger(), precedingMeasurementExists,
+            freeToBoundCorrection);
```

Comment:
might be outside this PR but I think material states before the first measurement are also not useful

Reply:
Right but we do not need to create a track state in this case technically. The covariance update can be independent from that.

Overall I think that is one more reason to keep all the track states and trim them afterwards to reduce all this special casing.